compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE DeriveAnyClass #-}
+module DA.Daml.LF.TypeChecker.Upgrade (checkUpgrade, Upgrading(..)) where
+
+import           Control.DeepSeq
+import           Control.Monad (unless, forM_)
+import           DA.Daml.LF.Ast as LF
+import           DA.Daml.LF.Ast.Recursive (TypeF(..))
+import           Data.Functor.Foldable (cata)
+import           Data.Foldable (fold)
+import           DA.Daml.LF.Ast.Alpha (alphaExpr, alphaType)
+import           DA.Daml.LF.TypeChecker.Env
+import           DA.Daml.LF.TypeChecker.Error
+import           Data.Data
+import           Data.Hashable
+import qualified Data.HashMap.Strict as HMS
+import qualified Data.NameMap as NM
+import qualified Data.Text as T
+import           Development.IDE.Types.Diagnostics
+import           GHC.Generics (Generic)
+
+data Upgrading a = Upgrading
+    { past :: a
+    , present :: a
+    }
+    deriving (Eq, Data, Generic, NFData, Show)
+
+instance Functor Upgrading where
+    fmap f Upgrading{..} = Upgrading (f past) (f present)
+
+instance Foldable Upgrading where
+    foldMap f Upgrading{..} = f past <> f present
+
+instance Traversable Upgrading where
+    traverse f Upgrading{..} = Upgrading <$> f past <*> f present
+
+instance Applicative Upgrading where
+    pure a = Upgrading a a
+    (<*>) f a = Upgrading { past = past f (past a), present = present f (present a) }
+
+foldU :: (a -> a -> b) -> Upgrading a -> b
+foldU f u = f (past u) (present u)
+
+checkUpgrade :: Version -> Upgrading LF.Package -> [Diagnostic]
+checkUpgrade version package =
+    let result =
+            runGamma
+                (initWorldSelf [] (present package))
+                version
+                (checkUpgradeM package)
+    in
+    case result of
+      Left err -> [toDiagnostic DsError err]
+      Right () -> []
+
+checkUpgradeM :: MonadGamma m => Upgrading LF.Package -> m ()
+checkUpgradeM package = do
+    (upgradedModules, _new) <- checkDeleted (EUpgradeMissingModule . NM.name) $ NM.toHashMap . packageModules <$> package
+    forM_ upgradedModules checkModule
+
+extractDelExistNew
+    :: (Eq k, Hashable k)
+    => Upgrading (HMS.HashMap k a)
+    -> (HMS.HashMap k a, HMS.HashMap k (Upgrading a), HMS.HashMap k a)
+extractDelExistNew Upgrading{..} =
+    ( past `HMS.difference` present
+    , HMS.intersectionWith Upgrading past present
+    , present `HMS.difference` past
+    )
+
+checkDeleted
+    :: (Eq k, Hashable k, MonadGamma m)
+    => (a -> Error)
+    -> Upgrading (HMS.HashMap k a)
+    -> m (HMS.HashMap k (Upgrading a), HMS.HashMap k a)
+checkDeleted handle upgrade = do
+    let (deleted, existing, new) = extractDelExistNew upgrade
+    case HMS.toList deleted of
+      ((_, head):_) -> throwWithContext $ handle head
```

Comment:
Would avoid overloading `head`

Reply:
@dylant-da I think you forgot to push or got blocked by the pre-commit checks